extension/js/common/api/key-server/wkd.ts

```diff
@@ -36,24 +40,42 @@ export class Wkd extends Api {
       // the proper fix would be to run encodeZBase32 through background scripts
       return { pubkey: null, pgpClient: null };
     }
-    const hu = opgp.util.encodeZBase32(await opgp.crypto.hash.digest(opgp.enums.hash.sha1, Buf.fromUtfStr(user)));
-    // todo - could also search on `https://openpgpkey.{domain}/.well-known/openpgpkey/{domain}/hu/{hu}?l={user}`
-    const url = `https://${recipientDomain}/.well-known/openpgpkey/hu/${hu}?l=${encodeURIComponent(user)}`;
+    const directDomain = recipientDomain.toLowerCase();
+    const advancedDomainPrefix = (directDomain === 'localhost') ? '' : 'openpgpkey.';
+    const hu = opgp.util.encodeZBase32(await opgp.crypto.hash.digest(opgp.enums.hash.sha1, Buf.fromUtfStr(user.toLowerCase())));
+    const directHost = (typeof this.port === 'undefined') ? directDomain : `${directDomain}:${this.port}`;
+    const advancedHost = `${advancedDomainPrefix}${directHost}`;
+    const userPart = `hu/${hu}?l=${encodeURIComponent(user)}`;
+    const advancedUrl = `${this.protocol}://${advancedHost}/.well-known/openpgpkey/${directDomain}`;
+    const directUrl = `${this.protocol}://${directHost}/.well-known/openpgpkey`;
     let binary: Buf;
+    let validUrl: string;
+    for (const url of [advancedUrl, directUrl]) {
+      try {
+        await Wkd.download(`${url}/policy`, undefined, 4);
+        console.info(`Policy found: ${url}/policy`);
+        validUrl = url;
+        break;
+      } catch (e) {
+        if (ApiErr.isNotFound(e) || ApiErr.isNetErr(e)) {
+          continue;
+        }
+        return { pubkey: null, pgpClient: null };
+      }
+    }
```

Comment:
@rrrooommmaaa  I think the error was here: if I'm reading it correctly, it was possible for the for loop to finish and `validUrl` would remain undefined. Then the lines below would continue calling that URL like `undefined/hu/...` causing the error reports.

I recommend to try to avoid for loops that do accounting in this manner unless one really cannot avoid them. It can be hard to spot bugs. I've rewritten it as a method that gets called two times explicitly with two different parameters (advanced vs direct url), which I think will make for an easier to debug structure.